pandas/io/excel.py

```diff
@@ -207,6 +210,18 @@ def __init__(self, io, **kwds):
         if engine is not None and engine != 'xlrd':
             raise ValueError("Unknown engine: %s" % engine)
 
+        # io can be a path object but relevant libraries may not be
+        # installed
+        # get_filepath_or_buffer() converts path objects to string
+        if _PATHLIB_INSTALLED:
+            from pathlib import Path
+            if isinstance(io, Path):
+                io = get_filepath_or_buffer(io)[0]
+        if _PY_PATH_INSTALLED:
+            from py.path import local as LocalPath
+            if isinstance(io, LocalPath):
```

Comment:
Calling `get_filepath_or_buffer()` on `io` no matter what almost works, except for URLs where it grabs the URL and returns the data as StringIO, the data needs to be kept as bytes I think. Is that what you were thinking or am I missing something? I'm struggling to find a way to run `io` through `get_filepath_or_buffer()` without messing up the URL case or losing track of what we originally had.  

The existing logic obviously works and we can keep it as long as we convert path objects to string before hitting that logic.